File: src/domain/eligibility.ts
import type { Claim, Conflict, ReportEligibility } from "./types";
import { WITHHOLDING_PRIVILEGE } from "./types";

/**
 * Compute a claim's report eligibility from its own state plus the state of
 * every conflict it participates in.
 *
 * Rule order is doctrine, not style:
 *  1. Privilege withholds — attorney-client / work-product / restricted
 *     material can never become report-eligible, whatever the review outcome.
 *  2. Rejected or superseded claims are excluded from reports (but remain
 *     visible in the record).
 *  3. Any unresolved conflict blocks report use.
 *  4. An escalated conflict parks the claim pending senior review.
 *  5. Only affirmatively supported or preserved claims are eligible.
 *  6. Everything else (unreviewed, disputed, escalated) still needs review.
 */
export function computeClaimEligibility(
  claim: Pick<Claim, "status" | "privilegeStatus" | "relatedConflictIds">,
  conflictsById: ReadonlyMap<string, Conflict>,
): ReportEligibility {
  if (WITHHOLDING_PRIVILEGE.includes(claim.privilegeStatus)) {
    return "privileged";
  }
  if (claim.status === "rejected" || claim.status === "superseded") {
    return "excluded";
  }
  const related = claim.relatedConflictIds
    .map((id) => conflictsById.get(id))
    .filter((c): c is Conflict => c !== undefined);
  if (related.some((c) => c.status === "unresolved")) {
    return "blocked_by_conflict";
  }
  if (related.some((c) => c.status === "escalated")) {
    return "needs_review";
  }
  if (claim.status === "supported" || claim.status === "preserved") {
    return "eligible";
  }
  return "needs_review";
}

File: server/src/atlas_argus/domain/eligibility.py
"""Report-eligibility rules. Mirrors ``src/domain/eligibility.ts``.

Rule order is doctrine, not style:
 1. Privilege withholds — attorney-client / work-product / restricted material
    can never become report-eligible, whatever the review outcome.
 2. Rejected or superseded claims are excluded from reports (but remain
    visible in the record).
 3. A quote nothing has confirmed cannot reach a report, whatever the review
    outcome — a reviewer approving a claim is not the same as the quote
    actually appearing in the evidence.
 4. Any unresolved conflict blocks report use.
 5. An escalated conflict parks the claim pending senior review.
 6. Only affirmatively supported or preserved claims are eligible.
 7. Everything else (unreviewed, disputed, escalated) still needs review.
"""

from __future__ import annotations

from collections.abc import Iterable

from .types import UNVERIFIED_QUOTE_STATES, WITHHOLDING_PRIVILEGE


def compute_claim_eligibility(
    claim_status: str,
    privilege_status: str,
    related_conflict_statuses: Iterable[str],
    quote_verification: str = "legacy_unverified",
) -> str:
    if privilege_status in WITHHOLDING_PRIVILEGE:
        return "privileged"
    if claim_status in ("rejected", "superseded"):
        return "excluded"
    # Legacy claims predate ingestion and are deliberately not blocked — doing
    # so would retroactively break every existing matter. They carry an honest
    # "unverified" label instead. Only a claim whose page was ingested and
    # could not confirm the quote is held back.
    if quote_verification in UNVERIFIED_QUOTE_STATES:
        return "needs_review"
    statuses = list(related_conflict_statuses)
    if "unresolved" in statuses:
        return "blocked_by_conflict"
    if "escalated" in statuses:
        return "needs_review"
    if claim_status in ("supported", "preserved"):
        return "eligible"
    return "needs_review"
